File: src/morphic/frontend/routes_inspector.py
"""
Inspector-tab API routes — EXIF scan, integrity check, edit, strip.
"""

from __future__ import annotations

import os
import time

from flask import Blueprint, jsonify, request

from morphic.inspector.scanner import get_job, start_job

bp = Blueprint("inspector", __name__)


# ── Scan (EXIF or Integrity) ──────────────────────────────────────────────


@bp.route("/scan", methods=["POST"])
def api_start_scan():
    """Start a new inspector scan (EXIF or integrity)."""
    data = request.get_json()
    if not data:
        return jsonify({"error": "No JSON body"}), 400

    folder = data.get("folder", "").strip()
    mode = data.get("mode", "exif")

    if not folder or not os.path.isdir(folder):
        return jsonify({"error": f"Invalid folder: {folder}"}), 400
    if mode not in ("exif", "integrity"):
        return jsonify({"error": "mode must be 'exif' or 'integrity'"}), 400

    job_id = start_job(folder=folder, mode=mode)
    return jsonify({"job_id": job_id}), 202


@bp.route("/scan/<job_id>/status")
def api_scan_status(job_id: str):
    """Get status of an inspector job."""
    job = get_job(job_id)
    if not job:
        return jsonify({"error": "Job not found"}), 404

    elapsed = 0.0
    if job.started_at:
        end = job.finished_at if job.finished_at else time.time()
        elapsed = end - job.started_at

    return jsonify({
        "id": job.id,
        "status": job.status,
        "mode": job.mode,
        "progress": job.progress,
        "message": job.message,
        "error": job.error,
        "total_files": job.total_files,
        "processed_files": job.processed_files,
        "elapsed_seconds": round(elapsed, 1),
    })


@bp.route("/scan/<job_id>/results")
def api_scan_results(job_id: str):
    """Get results of a completed inspector job."""
    job = get_job(job_id)
    if not job:
        return jsonify({"error": "Job not found"}), 404

    if job.status not in ("done", "error"):
        return jsonify({"error": "Scan not finished yet"}), 409

    return jsonify({
        "mode": job.mode,
        "results": job.results,
        "total_files": job.total_files,
    })


# ── EXIF Edit ──────────────────────────────────────────────────────────────


@bp.route("/exif/edit", methods=["POST"])
def api_exif_edit():
    """Edit EXIF fields on a single file."""
    from morphic.inspector.exif import edit_exif

    data = request.get_json()
    if not data:
        return jsonify({"error": "No JSON body"}), 400

    file_path = data.get("file", "").strip()
    updates = data.get("updates", {})

    if not file_path or not os.path.isfile(file_path):
        return jsonify({"error": f"Invalid file: {file_path}"}), 400
    if not updates:
        return jsonify({"error": "No updates provided"}), 400

    try:
        edit_exif(file_path, updates)
        return jsonify({"status": "ok", "file": file_path})
    except Exception as e:
        return jsonify({"error": str(e)}), 500


# ── EXIF Strip ─────────────────────────────────────────────────────────────


@bp.route("/exif/strip", methods=["POST"])
def api_exif_strip():
    """Strip EXIF from one or more files."""
    from morphic.inspector.exif import strip_exif_batch

    data = request.get_json()
    if not data:
        return jsonify({"error": "No JSON body"}), 400

    files = data.get("files", [])
    if not files:
        return jsonify({"error": "No files specified"}), 400

    results = strip_exif_batch(files)
    success_count = sum(1 for r in results.values() if r.get("success"))
    return jsonify({
        "results": results,
        "total": len(files),
        "success_count": success_count,
    })

File: web/routes_inspector.go
package web

import (
	"net/http"
	"os"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/exterex/morphic/internal/inspector"
)

func registerInspectorRoutes(r *gin.Engine) {
	g := r.Group("/api/inspector")
	{
		g.POST("/scan", handleInspectorScan)
		g.GET("/scan/:id/status", handleInspectorStatus)
		g.GET("/scan/:id/results", handleInspectorResults)
		g.POST("/exif/edit", handleExifEdit)
		g.POST("/exif/strip", handleExifStrip)
	}
}

func handleInspectorScan(c *gin.Context) {
	var req struct {
		Folder string `json:"folder"`
		Mode   string `json:"mode"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}
	if req.Folder == "" || !isDir(req.Folder) {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid folder: " + req.Folder})
		return
	}
	if req.Mode == "" {
		req.Mode = "exif"
	}
	if req.Mode != "exif" && req.Mode != "integrity" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "mode must be 'exif' or 'integrity'"})
		return
	}

	jobID := inspector.StartJob(req.Folder, req.Mode)
	c.JSON(http.StatusAccepted, gin.H{"job_id": jobID})
}

func handleInspectorStatus(c *gin.Context) {
	id := c.Param("id")
	job, ok := inspector.GetJob(id)
	if !ok {
		c.JSON(http.StatusNotFound, gin.H{"error": "Job not found"})
		return
	}

	elapsed := 0.0
	if !job.StartedAt.IsZero() {
		end := job.DoneAt
		if end.IsZero() {
			end = time.Now()
		}
		elapsed = end.Sub(job.StartedAt).Seconds()
	}

	c.JSON(http.StatusOK, gin.H{
		"id":              job.ID,
		"status":          job.Status,
		"mode":            job.Mode,
		"progress":        job.Progress,
		"message":         job.Message,
		"error":           job.Error,
		"total_files":     job.Total,
		"processed_files": job.Processed,
		"elapsed_seconds": round1(elapsed),
	})
}

func handleInspectorResults(c *gin.Context) {
	id := c.Param("id")
	job, ok := inspector.GetJob(id)
	if !ok {
		c.JSON(http.StatusNotFound, gin.H{"error": "Job not found"})
		return
	}

	if job.Status != "done" && job.Status != "failed" {
		c.JSON(http.StatusConflict, gin.H{"error": "Scan not finished yet"})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"mode":        job.Mode,
		"results":     job.Results,
		"total_files": job.Total,
	})
}

func handleExifEdit(c *gin.Context) {
	var req struct {
		File    string                 `json:"file"`
		Updates map[string]interface{} `json:"updates"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}
	if req.File == "" || !isFile(req.File) {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid file: " + req.File})
		return
	}
	if len(req.Updates) == 0 {
		c.JSON(http.StatusBadRequest, gin.H{"error": "No updates provided"})
		return
	}

	// EXIF editing with go-exif is read-only; for now return not-implemented
	// to match the API contract. A future iteration could use a subprocess.
	c.JSON(http.StatusNotImplemented, gin.H{
		"error": "EXIF editing is not yet supported in the Go version",
	})
}

func handleExifStrip(c *gin.Context) {
	var req struct {
		Files []string `json:"files"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}
	if len(req.Files) == 0 {
		c.JSON(http.StatusBadRequest, gin.H{"error": "No files specified"})
		return
	}

	results := inspector.StripExifBatch(req.Files)
	successCount := 0
	for _, r := range results {
		if s, ok := r["success"].(bool); ok && s {
			successCount++
		}
	}

	c.JSON(http.StatusOK, gin.H{
		"results":       results,
		"total":         len(req.Files),
		"success_count": successCount,
	})
}

func isFile(path string) bool {
	info, err := os.Stat(path)
	return err == nil && !info.IsDir()
}

func isDir(path string) bool {
	info, err := os.Stat(path)
	return err == nil && info.IsDir()
}

func round1(f float64) float64 {
	return float64(int(f*10+0.5)) / 10
}
